### PR TITLE
Bug 1415720 - Separate onboarding view from configuration. This will allow us to customize onboarding via LP.

### PR DESCRIPTION
**Pull request by farhanpatel**
The introViewController was very tightly coupled to our current intro slides. I've abstracted out the slides into `IntroCard`s that can be created on initialization of the ViewController. This will allow us to A/B test our onboarding. 

I haven't made the necessary LP changes. I will probably need to clarify with the MMA team first exactly what it is they want to test and how that will be represented in the LP backend.

### CODE REVIEW

**Comment by farhanpatel** [comment]
I've also made some changes to the text spacing on the slides. here is an iPhone X for example
- The title and description text now space out equally. 
- The description text no longer has custom spacing (this fixes some issues with text being cut off when the text is too long.)
Let me know what you think @tecgirl I can mess with the spacing some more if you like. 
Before
simulator screen shot - iphone x - 2018-01-09 at 09 48 59

after
simulator screen shot - iphone x - 2018-01-09 at 09 47 02

**Comment by tecgirl** [comment]
It would be great if we could match the previous padding specs in order to give comfortable space between the asset and title, lessen the whitespace gap below the title, and have more padding on the copy. 
ios-onboarding

File: Client/Frontend/Intro/IntroViewController.swift [hunk 1]
```diff
@@ -270,36 +195,50 @@ class IntroViewController: UIViewController, UIScrollViewDelegate {
...
-        return String(format: NSLocalizedString("Introductory slide %@ of %@", tableName: "Intro", comment: "String spoken by assistive technology (like VoiceOver) stating on which page of the intro wizard we currently are. E.g. Introductory slide 1 of 3"), NumberFormatter.localizedString(from: number, number: .decimal), NumberFormatter.localizedString(from: NSNumber(value: IntroViewControllerUX.NumberOfCards), number: .decimal))
+    func SELDynamicFontChanged(_ notification: Notification) {
+        guard notification.name == NotificationDynamicFontChanged else { return }
+        setupDynamicFonts()
     }
 
-    func changePage() {
-        let swipeCoordinate = CGFloat(pageControl.currentPage) * scrollView.frame.size.width
-        scrollView.setContentOffset(CGPoint(x: swipeCoordinate, y: 0), animated: true)
+    func DynamicFontChanged(_ notification: Notification) {
```
Comment (garvankeeley) [inline Client/Frontend/Intro/IntroViewController.swift]: Check SELDynamicFontChanged and DynamicFontChanged are both needed

File: Client/Frontend/Intro/IntroViewController.swift [hunk 2]
```diff
@@ -308,93 +247,122 @@ class IntroViewController: UIViewController, UIScrollViewDelegate {
 
     func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
         let page = Int(scrollView.contentOffset.x / scrollView.frame.size.width)
-        setActiveIntroView(introViews[page], forPage: page)
+        if let cardView = cardViews[safe: page] {
```
Comment (garvankeeley) [inline Client/Frontend/Intro/IntroViewController.swift]: cool subscript!

File: Client/Frontend/Intro/IntroViewController.swift [hunk 3]
```diff
@@ -308,93 +247,122 @@ class IntroViewController: UIViewController, UIScrollViewDelegate {
...
-    
-    fileprivate func setupDynamicFonts() {
-        startBrowsingButton.titleLabel?.font = UIFont(name: "FiraSans-Regular", size: DynamicFontHelper.defaultHelper.IntroStandardFontSize)
-        signInButton.titleLabel?.font = UIFont(name: "FiraSans-Regular", size: DynamicFontHelper.defaultHelper.IntroBigFontSize)
 
-        for titleLabel in titleLabels {
-            titleLabel.font = UIFont(name: "FiraSans-Medium", size: DynamicFontHelper.defaultHelper.IntroBigFontSize)
+    // Allows the scrollView to scroll while the CardView is in front
+    override func point(inside point: CGPoint, with event: UIEvent?) -> Bool {
+        if let buttonSV = button.superview {
```
Comment (garvankeeley) [inline Client/Frontend/Intro/IntroViewController.swift]: Nit: minor risk? perhaps the button could get nested in a future commit and the person not realize point(inside:with) will fail

Comment (farhanpatel) [reply Client/Frontend/Intro/IntroViewController.swift]: We were doing something similar before and at first its pretty odd why it exists. But when you miss around with it you realize very quickly why its there. The intro slides become impossible to scroll right away.

**Pull request merged**